Enable or disable in-home doorbell (via API or Homebridge Switch)

smockle opened the pull request:
# Details
Toggles the “Ring my in-home doorbell” setting (screenshot below):
![Ring my in-home doorbell](https://user-images.githubusercontent.com/3104489/66874829-73d68d00-ef7a-11e9-902c-03a11e650ca7.jpeg)

# Other Implementations
`python-ring-doorbell` implements this method in [`doorbot.py#L113-L133`](https://github.com/tchellomello/python-ring-doorbell/blob/6d2368b6615368a8724502d59755d14e0aa6e6f9/ring_doorbell/doorbot.py#L113-L133)

# Testing
I published [`homebridge-ring-smockle@5.8.8`](https://www.npmjs.com/package/homebridge-ring-smockle) to help with testing.

Note: If you toggle the switch in the Home app, then check status in the Ring app, you may see an outdated status, unless you’ve navigated to the main camera settings page.

# Example Use Cases
- Mute doorbell at night
- Mute doorbell when Do Not Disturb is enabled
- Mute doorbell when nursery white noise machine is turned on

dgreif commented on api/ring-types.ts:
```diff
@@ -295,26 +301,30 @@ export interface CameraData {
+    live_view_presets: string[]
+    motion_snooze_presets: string[]
+    doorbell_volume: number
+    chime_settings: {
```
Looks like this isn't defined on my stickup battery cam

```suggestion
    chime_settings?: {
```

smockle replied:
Thanks! Committed https://github.com/dgreif/ring/pull/152/commits/4789a682419e129cf5f382c18cb4e3ce84b2dee1 to fix.

dgreif commented on api/ring-camera.ts:
```diff
@@ -59,6 +73,7 @@ export class RingCamera {
       this.batteryLevel !== null &&
       this.batteryLevel < 100 &&
       this.batteryLevel >= 0)
+  existingDoorbellType = getExistingDoorbellType(this.initialData)
```
It would probably be best to change this to an observable so that we can subscribe to it in the homebridge layer and to prevent stale data being accessible via the api

```suggestion
  onExistingDoorbellType = this.onData.pipe(
    map(getExistingDoorbellType),
    distinctUntilChanged()
  )
```

smockle replied:
Makes sense! I implemented this incorrectly in https://github.com/dgreif/ring/pull/152/commits/a7fa4a2eeade1c2d0ef4454a0927b74371d126fd — after reading through the `lock.ts` code I see what I ought to have done. Working to update this PR now.

smockle replied:
Done—fixed in https://github.com/dgreif/ring/pull/152/commits/e13c4bb1a1ecff5f9cb71f539ca90348c241764c with a minor additional fix in https://github.com/dgreif/ring/pull/152/commits/08ce82871015b3df669703572bda9f253d33fbd1

dgreif commented on api/ring-camera.ts:
```diff
@@ -46,6 +47,19 @@ function getBatteryLevel(data: CameraData) {
+function getExistingDoorbellType(data: CameraData): DoorbellType | undefined {
+  if (
+    !data ||
+    !data.settings ||
```
From what I see in my devices, it looks like `data` and `data.settings` should always be defined.  Did you see any cases where they might not be?

smockle replied:
Nope—I have two cameras, and both have `data.settings`. I’ll remove the unnecessary checks, thanks!

smockle replied:
Fixed in https://github.com/dgreif/ring/pull/152/commits/20324e41d9e21933e39a36c4bd152fdddb62fd8f

dgreif commented on api/ring-types.ts:
```diff
@@ -295,26 +301,30 @@ export interface CameraData {
+    pir_motion_zones?: number[]
+    floodlight_settings?: any
+    light_schedule_settings?: any
+    luma_light_threshold?: number
```
Good catch with all these settings! Got em at the wrong level when I created the types I guess

dgreif commented on api/ring-camera.ts:
```diff
@@ -46,6 +47,19 @@ function getBatteryLevel(data: CameraData) {
   return batteryLevel
 }
 
+function getExistingDoorbellType(data: CameraData): DoorbellType | undefined {
```
What do you think about using a different name? `Existing` is a bit vague, but I'm not sure what would be better.  `Physical`, `Mechanical`, `InHome`, `HardWired`...any other ideas?

smockle replied:
I agree! Switched the name to `InHome`, since it matches the text in the Ring app UI (“Ring my in-home doorbell”): https://github.com/dgreif/ring/pull/152/commits/8db456b598f1616431dd0c1158d5cfd641a9fe05

smockle commented on api/ring-camera.ts:
```diff
@@ -156,6 +171,35 @@ export class RingCamera {
+          enable: on
+        }
+      }
+    })
```
Thanks! Fixed in https://github.com/dgreif/ring/pull/152/commits/a7fa4a2eeade1c2d0ef4454a0927b74371d126fd

smockle commented:
Thanks for the suggestions @dgreif; I appreciate you taking the time!

I’ve addressed the feedback from your initial review. I updated this PR and published a new version of the test package: [`homebridge-ring-smockle@5.8.8`](https://www.npmjs.com/package/homebridge-ring-smockle).

With the test package, when I toggle the “Ring my in-home doorbell” setting in the Ring app, I see the updated switch status reflected in the Home app, and vice versa.

dgreif commented:
Released in v`5.9.0`.  Thanks for the hard work on this, it worked great when I tested it with my doorbell.  Opens a world of possibilities for all those parents who keep putting "Don't Ring the Doorbell!" signs on their doors 😂
